Save Hints for Scripting

C-Loftus opened the pull request:
This pull request implements code for saving hints to a local map based on a syntactic name. These hints can then be used for writing scripts within Talon files.

This should probably not be merged yet, as I have to do some work cleaning it up a bit, but I wanted to get feedback on the general structure.

## Summary:
The user can then write scripts using this syntactic name as the target. All these saved hints are scoped to a particular website, and as such, it is up to the user to make sure that the browser is actually on that page when this script is written (this can be easily done within Talon).

There is generally very little change regarding the structure of the other code. The only thing I added that would change shared functions is the fact that `runRangoActionWithTarget` can optionally take in a wrapper if it has already been constructed (in this case, we saved the unique selector, so we want to use that unique selector and we can just construct the wrapper ourselves, and not use the hint name like other actions).

When the user saves an element, the inner text will change to the syntactic name and increase in size to show that it has been saved.

## Things I Am Unsure About:
To my understanding, we can currently only pass one argument in at a time (i.e., `request.arg` is only one item). As such, I am splitting on a string delimiter, but this seems a bit messy. I wasn't sure if there's a better way to do this or if we should leave it like this in order to preserve the fact that other functions just take one argument. I am still familiarizing myself with the TypeScript type system, and I am probably missing something on this though.

C-Loftus committed:
initial uniqueID code

C-Loftus committed:
Save hints and script on them

C-Loftus committed:
fix typos

C-Loftus committed:
fixed another mistakenly removed string

david-tejada commented:
I did a quick check and here are a couple things you need to change. I'll annotate more specific stuff when I have some time:
- Fix all those linting errors. Install XO to vscode so you can see them inline.
- Remove the function `styleSavedHints`. Use the function `showTooltip` for that purpose.

C-Loftus committed:
better error handling and delete mark function

C-Loftus commented:
Okay sounds good I will do that.  I just added some functionality to remove marks. I also switched any sort of error handling that should be visible to the user to use notify  and generally cleaned it up a bit. ( I think at least some of the linting errors are no longer in the most recent comment)

Do we have a function that can block until hints are loaded?  I have been able to test functionality with this and scripting works without issue, but  I think there are some instances in which it would be good to slow the script down on pages that are slow to load

david-tejada commented:
> Do we have a function that can block until hints are loaded? I have been able to test functionality with this and scripting works without issue, but I think there are some instances in which it would be good to slow the script down on pages that are slow to load

Not sure what you mean. Hints are never "finished" loading as they respond to the page adding more nodes, removing them, scrolling,... Maybe you want to wait till DOMContentLoaded, but I am not sure what you are trying to accomplish
https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event

C-Loftus committed:
Use tooltip fn

C-Loftus committed:
arg1 and arg2; cleaned up error edge case

C-Loftus commented:
Yes, I was thinking that if the user is writing a script that operates on multiple marks, they would have to wait until the first one is finished before triggering the second one. ( So they are in the proper context given the fact that the map doesn't store any fine grained context / dom state, but only the general website) We can easily use a simple sleep command in the Talon script for this, but I was wondering if it is possible to check this in the extension.

```
parse action for target
await page load
dispatch script
.... (run next action in talon script)
```

I believe using DOMContentLoaded would work, but I will double-check just to be sure

Additionally, I have made changes to the tooltip and modified the arguments to no longer require a string split.  No rush on reviewing this as always thanks for your input.

david-tejada commented:
You could wait until the element is present every time you want to perform an action on it. Similar to what I did [here for talon-filetree](https://github.com/paul-schaaf/talon-filetree/blob/cb801b558000196a974efc89f44ff3bde280fddb/src/fileExplorer.ts#L134-L157).

Another issue I would like you to address is the naming of this thing. In rango-talon you call them `marks`, here there are several other terms: `uniqueId`, `uniqueHint`, `savedId`. I personally would call them `references`. We can define a `reference` as key value pair that relates a word to an element/hint through it's unique css selector. I would change all the functions' names accordingly: `saveReference`, `removeReference`, `getReferencesForCurrentURL`, ...

I will give you feedback on the rest once you have committed these changes and fixed the linting errors.

C-Loftus committed:
Merge branch 'david-tejada:main' into save-hints-to-script

david-tejada committed:
Merge branch 'main' into c-loftus-references

david-tejada commented on src/typings/RangoAction.ts:
```diff
@@ -1,4 +1,5 @@
 interface RangoActionWithoutTargetWithoutArg {
+	arg: RangoActionWithTarget;
```
What is this doing here? Voice coding accident?

C-Loftus replied:
For us to script with an action we have  the scripting function and the function name that will be automatically run passed into it. So I think that this arg was referring to the specific Rango action that will operate on the reference (called a target in this instance)

That being said it might start to get confusing to have an action interface that takes an action as well so it maybe should be renamed to differentiate it from the other actions.

david-tejada commented on src/typings/RangoAction.ts:
```diff
@@ -94,6 +101,7 @@ interface RangoActionWithoutTargetWithOptionalNumberArg {
 }
 
 interface RangoActionWithTargets {
+	arg: any;
```
Another weird arg added

C-Loftus replied:
I was trying to make it accept any sort of Rango action so I thought that we need to have an argument that could work with any rango action. ie be either a distance to scroll or an element or something else like that, 

Now that we have coordinated that we are only going to accept a small subset of actions to script with it probably doesn't make sense to have this any more

C-Loftus replied:
Although now that I'm looking at it again I'm actually not even sure why this is here. I may have accidentally added this here instead of where the interface should be for the scripting actions

david-tejada replied:
Definitely not necessary. The only thing necessary is declaring an interface like this:

```typescript
interface RangoActionRunActionOnReference {
	type: "runActionOnReference";
	arg: RangoActionWithTarget["type"];
	arg2: string;
}
```

The only thing that might be missing here is another optional argument in case we want to scroll relative to a target with a custom scroll amount, but I don't see much use for that

david-tejada commented on src/background/setup/initBackgroundScript.ts:
```diff
@@ -103,6 +103,10 @@ export async function initBackgroundScript() {
+			await storeIfUndefined(
+				"savedIDsByHost",
+				new Map<string, Map<string, string>>()
+			);
```
This wasn't necessary, but this is on me as the previous two are also not necessary.

david-tejada commented on src/content/actions/actOnUniqueIDs.ts:
```diff
@@ -0,0 +1,171 @@
+}
+
+export async function rangoActionOnSavedID(
+	action: string,
```
You can do this in typescript to get the type of a property:

```typescript
type: RangoActionWithTarget["type"]
```

david-tejada committed:
Fix linting and type errors and several other improvements

david-tejada committed:
Create function withHostReferences to simplify getting and saving references

david-tejada committed:
Fix error showing tooltip when the wrapper doesn't have a hint

david-tejada committed:
Wait up to one second if the element with the unique selector isn't available

david-tejada commented:
Ok, I think this is done. Give it a try and let me know what you think.

david-tejada committed:
Fix references implementation to work with iframes

david-tejada commented:
Ok, now it works with iframes. I haven't done anything about the possibility of the same selector being found on different iframes. I think this is a remote possibility and it only showed up in that mdn page because they were just examples and the iframe only contained one element. In any case we can't be 100% sure the selectors will be totally unique even without taking into account iframes. A unique selector at some point could be not unique when the page is in a different state. A way to prevent this could be using another library that creates a more complex selector. This is a double edged sword as a minimal change in the page could break our references.

I have added a toast notification apart from the tooltip when deleting references as sometimes the element that is referenced might not be in the dom or visible in the viewport.

Give it a try and let me know if something doesn't work properly.

C-Loftus commented:
Microsoft teams is working as expected and dynamic content seems to work well.  Great job figuring this out I wouldn't have known how to do this 

However, I have encountered a few instances where creating the mark takes too long, resulting in a timeout and triggering a toast notification timeout error. Nonetheless, the mark is still saved on the extension side and can be scripted on as intended. 
It is totally anecdotal but the only time I've see this come up is  relatively shortly after startup so I wasn't sure if something is loading in the background and causing things to take a bit longer. ( it was on a relatively bloated New York times page)

david-tejada commented:
You could try reducing the number of maxCombinations and see if that helps

C-Loftus commented:
I'm not able to reproduce this error regardless whether or not I change the number of max combinations. Maybe it was some sort of quirk of my own setup, hard to say. 

Regardless I think this looks good now and pretty much all set. The only thing I'm thinking of is that maybe we should let the user know in some way about the potential for non unique iframes. I don't think anyone is going to be doing scripting on particularly high stakes tasks with this (banking, file management, etcetera) but I think it is a good practice just in case they do. I'm not sure where the best place to put that information would be . Perhaps just the read me is sufficient

david-tejada commented:
Sure, I will add a note. Not that I haven't clicked on the wrong thing a million times by now anyway when talon misunderstands me
